tsl/src/fdw/scan_plan.c

```diff
@@ -942,46 +1000,505 @@ add_foreign_grouping_paths(PlannerInfo *root, RelOptInfo *input_rel, RelOptInfo
+	 * the underlying grouping relation to perform the final sort remotely,
+	 * which is stored into the fdw_private list of the resulting path.
+	 */
+
+	/* Assess if it is safe to push down the final sort */
+	foreach (lc, root->sort_pathkeys)
+	{
+		PathKey *pathkey = (PathKey *) lfirst(lc);
+		EquivalenceClass *pathkey_ec = pathkey->pk_eclass;
+		Expr *sort_expr;
+
+		/*
+		 * is_foreign_expr would detect volatile expressions as well, but
+		 * checking ec_has_volatile here saves some cycles.
+		 */
+		if (pathkey_ec->ec_has_volatile)
+			return;
+
+		/* Get the sort expression for the pathkey_ec */
+		sort_expr = find_em_expr_for_input_target(root, pathkey_ec, input_rel->reltarget);
+
+		/* If it's unsafe to remote, we cannot push down the final sort */
+		if (!is_foreign_expr(root, input_rel, sort_expr))
+			return;
+	}
+
+	/* Safe to push down */
+	fpinfo->pushdown_safe = true;
+
+	/* Construct TsFdwPathExtraData */
+	fpextra = (TsFdwPathExtraData *) palloc0(sizeof(TsFdwPathExtraData));
+	fpextra->target = root->upper_targets[UPPERREL_ORDERED];
+	fpextra->has_final_sort = true;
+
+	/* Estimate the costs of performing the final sort remotely */
+	fdw_estimate_path_cost_size(root,
+								input_rel,
+								root->sort_pathkeys,
+								&rows,
+								&width,
+								&startup_cost,
+								&total_cost);
+
+	/*
+	 * Build the fdw_private list that will be used by GetForeignPlan.
+	 * Items in the list must match order in enum FdwPathPrivateIndex.
+	 */
+	fdw_private = list_make2(makeInteger(true), makeInteger(false));
+
+	/* Create foreign ordering path */
+	ordered_path = (Path *) create_path(root,
+										input_rel,
+										root->upper_targets[UPPERREL_ORDERED],
+										rows,
+										startup_cost,
+										total_cost,
+										root->sort_pathkeys,
+										NULL, /* no extra plan */
+										fdw_private);
+
+	/* and add it to the ordered_rel */
+	add_path(ordered_rel, (Path *) ordered_path);
+}
+
+/*
+ * add_foreign_final_paths
+ *		Add foreign paths for performing the final processing remotely.
+ *
+ * Given input_rel contains the source-data Paths.  The paths are added to the
+ * given final_rel.
+ */
+static void
+add_foreign_final_paths(PlannerInfo *root, RelOptInfo *input_rel, RelOptInfo *final_rel,
+						CreateUpperPathFunc create_path, FinalPathExtraData *extra)
+{
+	Query *parse = root->parse;
+	TsFdwRelInfo *ifpinfo = fdw_relinfo_get(input_rel);
+	TsFdwRelInfo *fpinfo = fdw_relinfo_get(final_rel);
+	bool has_final_sort = false;
+	List *pathkeys = NIL;
+	TsFdwPathExtraData *fpextra;
+	double rows;
+	int width;
+	Cost startup_cost;
+	Cost total_cost;
+	List *fdw_private;
+	Path *final_path;
+
+	/*
+	 * Currently, we only support this for SELECT commands
+	 */
+	if (parse->commandType != CMD_SELECT)
+		return;
+
+	/*
+	 * No work if there is no FOR UPDATE/SHARE clause and if there is no need
+	 * to add a LIMIT node
+	 */
+	if (!parse->rowMarks && !extra->limit_needed)
+		return;
```

Comment:
Looks like this function always exits here or statement above. The code below is never executed.